docs/_includes/Refinement.md

```diff
@@ -0,0 +1,130 @@
+# 23. Refinement {#sec-module-refinement}
+
+Refinement is the process of replacing something somewhat abstract with something somewhat more concrete.
+For example, in one module one might declare a type name, with no definition,
+such as `type T`, and then in a refining module, provide a definition. 
+One could prove general properties about the contents of an (abstract) module,
+and use that abstract module, and then later provide a more concrete implementation without having to redo all of the proofs.
+
+In Dafny, refinement takes the following form. One module
+declares some program entities. A second module _refines_ the first by
+declaring how to augment or replace (some of) those program entities.
+The first module is called the _refinement parent_; the second is the
+_refining_ module; the result of combining the two (the original declarations
+and the augmentation instructions) is the _combined_ module or _refinement result_.
+
+Syntactically, the refinement parent is a normal module declaration.
+The refining module declares which module is its refinement parent with the
+`refines` clause:
+```
+module P { // refinement parent
+}
+module M refines P { // refining module
+}
+```
+
+The refinement result is created as follows.
+
+0) The refinement result is a module within the same enclosing module as the
+refining module, has the same name, and in fact replaces the refining module  in their shared scope.
```

Comment:
I see. The description talk about three things (refinement parent, refining module, refinement result), not just two (`P` and `M`). I like it.

Optionally, this text could add the remark that the "refining module" is like a set of directives that will make the refinement result from the refinement parent, almost like a fancy macro.

Reply:
Added